### PR TITLE
[4.0] RavenDB-11396 Better error for wrong host name

### PR DESCRIPTION
**Pull request by iftahbe**

### CODE REVIEW

File: src/Raven.Client/Http/RequestExecutor.cs [hunk 1]
```diff
@@ -1396,6 +1402,25 @@ internal static bool OnServerCertificateCustomValidationCallback(object sender,
             return false;
         }
 
+        private static void ThrowNameMismatchException(object sender, X509Certificate cert)
+        {
+            var hostname = ((HttpRequestMessage)sender).RequestUri.DnsSafeHost;
+            var cert2 = new X509Certificate2(cert.Handle);
```
Comment (iftahbe) [inline src/Raven.Client/Http/RequestExecutor.cs]: @ayende This is a difference between 4.0 and 4.1
Netstandard1.3 doesn't have a ctor which accepts x509certificate so using the handle instead.

File: src/Raven.Client/Exceptions/Security/NameMismatchException.cs [hunk 2]
```diff
@@ -0,0 +1,19 @@
+﻿using System;
+
+namespace Raven.Client.Exceptions.Security
+{
+    public class NameMismatchException : AuthenticationException
```
Comment (ppekrol) [inline src/Raven.Client/Exceptions/Security/NameMismatchException.cs]: CertificateNameMismatchException maybe?

File: src/Raven.Client/Http/RequestExecutor.cs [hunk 3]
```diff
@@ -1396,6 +1402,29 @@ internal static bool OnServerCertificateCustomValidationCallback(object sender,
             return false;
         }
 
+        private static void ThrowNameMismatchException(object sender, X509Certificate cert)
+        {
+            var hostname = ((HttpRequestMessage)sender).RequestUri.DnsSafeHost;
```
Comment (ppekrol) [inline src/Raven.Client/Http/RequestExecutor.cs]: It is not always HttpRequestMessage. We are using this callback also in Changes API with WebSockets. Then the sender is SslStream

File: src/Raven.Client/Http/RequestExecutor.cs [hunk 4]
```diff
@@ -1396,6 +1402,29 @@ internal static bool OnServerCertificateCustomValidationCallback(object sender,
             return false;
```
Comment (ppekrol) [inline src/Raven.Client/Http/RequestExecutor.cs:1407]: Should we also check and throw here?

File: src/Raven.Client/Http/RequestExecutor.cs [hunk 5]
```diff
@@ -1396,6 +1402,29 @@ internal static bool OnServerCertificateCustomValidationCallback(object sender,
...
         }
 
+        private static void ThrowNameMismatchException(object sender, X509Certificate cert)
+        {
+            var hostname = ((HttpRequestMessage)sender).RequestUri.DnsSafeHost;
+#if NETSTANDARD1_3
+            var cert2 = new X509Certificate2(cert.Handle);
+#else
+            var cert2 = new X509Certificate2(cert);
```
Comment (ppekrol) [inline src/Raven.Client/Http/RequestExecutor.cs]: var cert2 = cert as X509Certificate2 ?? new X509Certificate2(cert);

File: src/Raven.Client/Http/RequestExecutor.cs [hunk 6]
```diff
@@ -1396,6 +1402,29 @@ internal static bool OnServerCertificateCustomValidationCallback(object sender,
             return false;
         }
 
+        private static void ThrowNameMismatchException(object sender, X509Certificate cert)
+        {
+            var hostname = ((HttpRequestMessage)sender).RequestUri.DnsSafeHost;
+#if NETSTANDARD1_3
+            var cert2 = new X509Certificate2(cert.Handle);
```
Comment (ppekrol) [inline src/Raven.Client/Http/RequestExecutor.cs]: var cert2 = cert as X509Certificate2 ?? new X509Certificate2(cert.Handle);

It looks like not much is happening in ctor, but let's do this.

**Review by ppekrol — CHANGES_REQUESTED** [review]

File: src/Raven.Client/Http/RequestExecutor.cs (hunk 4 above)
Comment (ayende) [reply src/Raven.Client/Http/RequestExecutor.cs:1407]: I think we better. The major issue is that if we don't, we get a pretty cryptic (pun intended) error.

File: src/Raven.Client/Http/RequestExecutor.cs (hunk 5 above)
Comment (ayende) [reply src/Raven.Client/Http/RequestExecutor.cs]: We already have code that does this, IIRC

**Review by ppekrol — APPROVED** [review]

**Pull request merged**